### ISSUE TITLE
Swarm: Network dropping packets? Way to debug networking?

### ISSUE THREAD

**Issue by narcoticfresh** [moby/moby#34136]
**Preamble**

First: This is *maybe* a bug of docker, but we are not sure. We are experiencing a strange behavior (over at vfarcic/docker-flow-proxy#287) of two battle tested components (haproxy and nginx) in the context of a simple local swarm cluster and we are not able debug the problems any further. So we are reaching out to this project for helping us find the cause of the problem.

**Description**

Given a simple swarm cluster containing two services (haproxy with a generated config based on metadata and nginx) and haproxy proxying nginx using `tcpmode`, it seems connections are dropped or cannot be established. 

Having nginx and haproxy running and executing 200 requests in a bash script, many requests (from the client side perspective) return no or errornous response. 

**Important point**: The nginx in question is the bare default nginx image and has no logic - it just serves the default welcome page. I switched the workload with an Apache container, this made no difference.

**Steps to reproduce the issue:**

[I prepared a gist](https://gist.github.com/narcoticfresh/ec2ff31ae2c96bd5160f95de85141c35) containing everything needed to reproduce it. 

(assuming you have swarm cluster running that you can deploy to)
1. [Fetch this compose file](https://gist.github.com/narcoticfresh/ec2ff31ae2c96bd5160f95de85141c35#file-stack-proxy-test-yml) and deploy it to your cluster
2. Once all is running, [get this bash script](https://gist.github.com/narcoticfresh/ec2ff31ae2c96bd5160f95de85141c35#file-test-req-sh), replace the IP there with your manager node ip and run it
3. Check the `docker logs` of the `proxy` service in the stack and look out for lines like: `tcpFE_88 tcpFE_88/<NOSRV>` - those were requests haproxy could not serve as it didn't find an upstream server.
4. You can check the file `/cfg/haproxy.cfg` in the `proxy` container and see that this is a valid haproxy configuration.

**Describe the results you received:**

I see two battle tested components (haproxy and nginx), both having valid configurations failing to interact properly from the client side perspective by the fact that haproxy is not able to serve 200 requests to the client side.

**Describe the results you expected:**

After firing the 200 requests via curl, I expect to not see errors in the haproxy log and having received 200 valid responses on the client side.

**Additional information you deem important (e.g. issue happens only occasionally):**

* This example was executed on a local swarm cluster having 3 nodes (1 manager, 2 worker) provisioned with `docker-machine`, no AWS or whatever involved ;-)
* To rule out that `docker-flow-proxy` is the culprit, I moved the generated configuration to the official `haproxy` container, tried to fiddle around with other timeouts and other settings - to no use ;-/
* *We are having no problems when we configure haproxy to use `mode http` to the upstream service - only on `mode tcp` we are observing this behavior*

:point_right: I realize that this is a strange problem. As said, I am merely looking for insights/ideas how to debug and further analyze this problem if this could not be a docker networking problem.

It just looks like a possible bug in swarm networking(?)

**Output of `docker version`:**

```
Client:
 Version:      17.05.0-ce
 API version:  1.29
 Go version:   go1.7.5
 Git commit:   89658be
 Built:        Thu May  4 22:10:54 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.05.0-ce
 API version:  1.29 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   89658be
 Built:        Thu May  4 21:43:09 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

**Output of `docker info`:**

```
Containers: 9
 Running: 6
 Paused: 0
 Stopped: 3
Images: 29
Server Version: 17.05.0-ce
Storage Driver: aufs
 Root Dir: /mnt/sda1/var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 155
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: bridge host macvlan null overlay
Swarm: active
 NodeID: v9paxd8atn5dv0xae7zzyoio4
 Is Manager: true
 ClusterID: vkhyiz7q3q0l81sozylkfd481
 Managers: 1
 Nodes: 3
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Number of Old Snapshots to Retain: 0
  Heartbeat Tick: 1
  Election Tick: 3
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
 Node Address: 192.168.99.102
 Manager Addresses:
  192.168.99.102:2377
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 9048e5e50717ea4497b757314bad98ea3763c145
runc version: 9c2d8d184e5da67c95d601382adf14862e4f2228
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.4.66-boot2docker
Operating System: Boot2Docker 17.05.0-ce (TCL 7.2); HEAD : 5ed2840 - Fri May  5 21:04:09 UTC 2017
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 995.8MiB
Name: node0
ID: Q3RQ:QS5U:QCHG:4DBG:BI2J:3EI7:4TTB:UPHU:A5GQ:KN2E:COVB:QOTZ
Docker Root Dir: /mnt/sda1/var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 136
 Goroutines: 285
 System Time: 2017-07-17T08:54:13.027299533Z
 EventsListeners: 6
Username: narcoticfresh
Registry: https://index.docker.io/v1/
Labels:
 provider=virtualbox
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

This was tested (and IMHO can reproduced) on a `docker-machine` provisioned swarm cluster using virtualbox and default settings.

**Comment by narcoticfresh** [comment]
Hm, I've seen #32195 *before* opening this issue but thought that it only relates to more complex AWS setups, so I thought it's not applicable for our simple case.

Now i've seen docker 17.06 is out, I'll try to upgrade and see if this betters things..

**Comment by jefflill** [comment]
I'm seeing similar problems:

* Running 17.06.0-ce on Ubuntu 16.04
* HAProxy is running as a service on **all** nodes proxying both TCP and HTTP endpoints from ports published to the ingress mesh network.
* The TCP and HTTP endpoints forward to the same backend HTTP service.

For me, the problem is submitting requests from an external Windows 10 machine to the HAProxy HTTP endpoint.  Running a simple script results in maybe 1/2 to 3/4s of the requests failing because a connection cannot be established.  Running tcpdump on a Docker host (HAProxy is running on all of them), shows the SYN packet coming in from Windows but there isn't an ACK.

The strange thing for me is that the TCP HAProxy endpoints seem to be working fine, even though they're hitting the same Docker node and are being routed to the same backend service.  I can also successfully submit requests from Linux machines that belong and also do not belong to the cluster.

This is very frustrating.  I've been struggling with this for several days now.  I wonder if mesh networking actually works at all.

**Is anybody actually using this in production?**

I went back to 17.03.0-ce and am seeing the same thing.  Perhaps 1.13 works, I'll try that next.

**Comment by SFedyukin** [comment]
Hello!
I have crazy week with similar problem.
Solved with kernel 4.12.
Docker swarm does not support kernel 3.x???

**Comment by thaJeztah** [comment]
Let me close this ticket for now, as it looks like it went stale.
